### PR TITLE
[Http Instrumentation] Unify exposed public API

### LINKED ISSUE #3434: [HttpClient instrumentation] Runtime failures with a mix of NetFramework and NetStandard
# Bug Report

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):
OpenTelemetry.Instrumentation.Http: 1.0.0-rc9.4"

Runtime version (e.g. `net462`, `net48`, `netcoreapp3.1`, `net6.0` etc. You can
find this information from the `*.csproj` file):

"Common component": netstandard2.0
"NetFramework frontent": net472
"NetCore frontend": net5.0

## Symptom

My application runs on both Net Framework and Net 5.0 as we are migrating between frameworks.
We have a common netstandard2.0 project shared between the two, to minimize code duplication.

Our OTel tracing setup happens in the common dll that is compiled with netstandard2.0, but when I run the net472 service that incorporates this dll, I get the following exception:

Unhandled Exception: Method not found:
'OpenTelemetry.Trace.TracerProviderBuilder OpenTelemetry.Trace.TracerProviderBuilderExtensions.AddHttpClientInstrumentation(OpenTelemetry.Trace.TracerProviderBuilder, System.Action`1<OpenTelemetry.Instrumentation.Http.HttpClientInstrumentationOptions>)'. at MyServiceNamespace.CommonLib.LoggingSetup.AddCommonInstrumentation(TracerProviderBuilder instance)

My understanding is that this is happening because my common library compiled against the netstandard version of OpenTelemetry.Instrumentation.Http, but at runtime, the .Net Framework version gets packaged with my app, and the .Net Framework version of the library has a slightly different method signature.

**What is the expected behavior?**

I expect that methods packed into the netstandard2.0 assembly will also be present in netframework and netcore assemblies.

**What is the actual behavior?**

A method with a different signature is available in the netframework assembly, causing runtime errors.

## Reproduce

Fairly self-expalanatory.

## Additional Context

The problem originates [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/85ba904b48ac7a8315b6adaf8c1e7a29051ce672/src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs#L31). I wonder if the modern path ("else" case) should be present on both runtimes. Alternatively, this nuget should not target netstandard2.0, and instead target net4xx and netstandard2.1, to avoid this unreliable behavior.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #3434 (really mean it this time)
Relates to #3787

## Changes

* Exposes the same API for all targets to fix package resolution issues when consuming Http instrumentation into `netstandard2.0` projects.

* We used to only run HttpWebRequest tests for .NET Framework & HttpClient tests for .NET/Core. Now we run the full suite for all runtimes.

## Public API Changes

```diff
namespace OpenTelemetry.Instrumentation.Http
{
-#if NETFRAMEWORK
-    public class HttpWebRequestInstrumentationOptions
-    {
-        public Func<HttpWebRequest, bool> Filter { get; set; }
-        // omitted other stuff but whole class removed
-    }
-#else
    public class HttpClientInstrumentationOptions
    {
-        public Func<HttpRequestMessage, bool> Filter { get; set; }
+        public Func<HttpRequestMessage, bool> FilterHttpRequestMessage { get; set; }
+        public Func<HttpWebRequest, bool> FilterHttpWebRequest { get; set; }
    }
-#endif
}
```

Not shown: Removed the `AddHttpClientInstrumentation` extensions using `HttpWebRequestInstrumentationOptions` and made the ones using `HttpClientInstrumentationOptions` available to `NETFRAMEWORK` targets.

## TODOs

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [X] Unit tests
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs [hunk 1]
```diff
@@ -28,12 +29,48 @@ namespace OpenTelemetry.Instrumentation.Http
...
+        /// <item>The return value for the filter function is interpreted as:
+        /// <list type="bullet">
+        /// <item>If filter returns <see langword="true" />, the request is
+        /// collected.</item>
+        /// <item>If filter returns <see langword="false" /> or throws an
+        /// exception the request is NOT collected.</item>
+        /// </list></item>
+        /// </list>
+        /// </remarks>
+        public Func<HttpWebRequest, bool> FilterHttpWebRequest { get; set; }
```
Comment (alanwest) [inline src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs:73]: Would this be useful to not pollute the API for net6.0 folk?

```suggestion
#if NETFRAMEWORK || NETSTANDARD2_0
        public Func<HttpWebRequest, bool> FilterHttpWebRequest { get; set; }
#endif
```

Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.Http/HttpClientInstrumentationOptions.cs:73]: Ohh... I just realized the `net6.0` target was removed. Is it possible to keep it?

**Comment by CodeBlanch** [comment]
@vishweshbankwar FYI I updated this to include #3792.

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.Http/README.md [hunk 2]
```diff
@@ -72,30 +72,32 @@ For an ASP.NET application, adding instrumentation is typically done in the
...
-`HttpWebRequestInstrumentationOptions` (.NET Framework applications). It is
-important to note that even if `HttpClient` is used in .NET Framework
-applications, it underneath uses `HttpWebRequest`. Because of this,
-`HttpWebRequestInstrumentationOptions` is the configuration option for .NET
-Framework applications, irrespective of whether `HttpWebRequest` or `HttpClient`
-is used.
+`HttpClientInstrumentationOptions`. It is important to note that there are
+differences between .NET Framework and newer .NET/.NET Core runtimes which
+govern what options are used. On .NET Framework, `HttpClient` uses the
+`HttpWebRequest` API. On .NET & .NET Core, `HttpWebRequest` uses the
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/README.md:78]:
```suggestion
`HttpWebRequest` API. On .NET & .NET Core, `HttpClient` uses the
```

**Review by cijothomas — APPROVED** [review]

**Review by vishweshbankwar — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.Http/README.md (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.Http/README.md:78]: yea i can see this is not easy to articulate. Lets merge this now, and come back to fixing the wording.

**Pull request merged**